gr-digital/lib/mimo_channel_estimator_cc_impl.cc

```diff
@@ -0,0 +1,246 @@
+#endif
+        }
+      }
+    }
+
+    int
+    mimo_channel_estimator_cc_impl::general_work (int noutput_items,
+                       gr_vector_int &ninput_items,
+                       gr_vector_const_void_star &input_items,
+                       gr_vector_void_star &output_items)
+    {
+      // Collect all tags of the input buffer with key "csi" in the vector 'tags'.
+      get_tags_in_window(tags, 0, 0, noutput_items, d_key);
+
+      uint32_t nconsumed = 0; // Number of read items.
+      uint32_t nwritten = 0; // Number of written items.
+      uint32_t symbol_length; // Number of items in the current symbol.
+
+      if(tags.size() == 0){ // Input buffer includes no tags at all.
+        // Handle all samples in buffer as they belong to the current symbol.
+        symbol_length = noutput_items;
+        // Copy the data to the output buffer.
+        copy_symbols(input_items, output_items, symbol_length, 0, 0);
+        nconsumed = noutput_items;
+        nwritten = noutput_items;
+      } else { // Input buffer includes tags.
+        if (tags[0].offset - nitems_read(0) > 0){
+          /* There are items in the input buffer, before the first tag arrives,
++          * which belong to the previous symbol. */
+          symbol_length = tags[0].offset - nitems_read(0);
+          // Copy the data to the output buffer.
+          copy_symbols(input_items, output_items, symbol_length, 0, 0);
+          nconsumed += symbol_length;
+          nwritten += symbol_length;
+        }
+        // Iterate over tags in buffer.
+        for (unsigned int i = 0; i < tags.size(); ++i) {
+          // Calculate the number of items before the next tag.
+          if (i < tags.size() - 1) {
+            // This is not the last tag in the buffer.
+            symbol_length = tags[i + 1].offset - nitems_read(0) - nconsumed;
+            // Check if the symbol length is smaller than the training length
+            if (symbol_length < d_training_length){
+              // We ignore this unfinished training sequence and dump it without any estimation.
+              nconsumed += symbol_length;
+              GR_LOG_INFO(d_logger, format("The symbol length (%d) is smaller than the training length (%d)."
+                                           "No estimation possible. Dumping symbol.")
+                                    %symbol_length %d_training_length);
+              continue;
+            }
+          } else {
+            // This is the last tag in the buffer.
+            symbol_length = noutput_items - nconsumed;
+            // Check if the training sequence of this last symbol is completely in this buffer.
+            if (symbol_length < d_training_length){
+              /* We consume the whole training sequence in the next buffer and don't consume it now
+               * in order to do the estimation at once. */
+              break;
```

Comment:
this is the last iteration of the `for` loop anyway, and this is the last statement in that iteration. Breaking at this point doesn't _do_ anything, or am I mistaken?

Reply:
are we perhaps missing the case where symbol length is _not_ smaller than training sequence length and we need to consume now?

Reply:
The for loop is not finished after the if else statement. If the symbol length is smaller than the training length, we cannot process the received training sequence right now. So we are breaking the loop (we are on the last iteration, but there would still be the copy and tag assignment operations in this iteration to be done) to skip the rest of the code in the for loop.

Reply:
Wait, I'll have to talk to you personally about this, but this is the "else" belonging to the "if(not last iteration)" clause, and that means this else is only executed on the last iteration, and right after the closing bracket of the "else" clause comes the closing bracket of the "for" clause. I'm clearly missing something here!